src/plan/tracing.rs

```diff
@@ -1,104 +1,111 @@
 //! This module contains code useful for tracing,
 //! i.e. visiting the reachable objects by traversing all or part of an object graph.
 
-use std::mem;
-
 use crate::scheduler::gc_work::{EdgeOf, ProcessEdgesWork};
 use crate::scheduler::{GCWorker, WorkBucketStage};
 use crate::util::ObjectReference;
 use crate::vm::EdgeVisitor;
 
 /// This trait represents an object queue to enqueue objects during tracing.
-pub trait ObjectQueue {
+pub trait Queue<T> {
```

Comment:
We may give it a better name, so we don't confuse it with a common data structure, such as `VecDeque`.

Also, my original intention of turning `TransitiveClosure` into an `ObjectQueue` is **to give `trace_object` a place to enqueue nodes when an object is first visited**.  We may preserve a trait for that purpose.  May be call it `FirstVisitedObjectCallback`, or change it back to `ProcessNodeOfTransitiveClosure`, or `ObjectVisitor`, or something simpler?

And we may create a different trait for a "linear data structure with a given capacity".  I think that's what you want to generalise, so that the limited-capacity structure is not limited to `ObjectReference`, but can also hold `MemorySlice` as well.

Then we let the "linear data structure with a given capacity" implement the "place to put objects when `trace_object` first visits an object".  By doing this, we make it clear that they are two different purposes, but the former happens to be a good sink for the latter in `ProcessEdgesWork`.  But when we do LXR in concurrent marking, we directly process first-visited objects instead of putting them into a "linear data structure with a given capacity".

Reply:
This is reverted. I still use `VectorQueue<T>` for the modbufs.